package edu.mayo.ve.VCFParser;

import com.mongodb.BasicDBObject;
import com.mongodb.DBObject;
import com.mongodb.util.JSON;
import com.tinkerpop.pipes.Pipe;
import edu.mayo.TypeAhead.TypeAheadCollection;
import edu.mayo.pipes.bioinformatics.VCF2VariantPipe;
import edu.mayo.pipes.history.History;
import edu.mayo.senders.FileSender;
import edu.mayo.senders.Sender;
import org.junit.Ignore;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

/**
 * Created with IntelliJ IDEA.
 * User: m102417
 * Date: 1/14/14
 * Time: 11:00 AM
 * VCFParser tests take a lot of time, so make sure you look at the VCFParserITCase for the tests that run slower
 */
public class VCFParserTest {

    @Ignore("depricated") @Test
    public void testRemoveSamples(){
        String input =  "{ \"CHROM\" : \"chr1\" , \"POS\" : \"61871\" , \"ID\" : \".\" , \"REF\" : \"C\" , \"ALT\" : \"CT\" , \"QUAL\" : \".\" , \"FILTER\" : \"PASS\" , \"INFO\" : { \"AC\" : [ 1] , \"AF\" : [  0.5], \"AN\" : 2 , \"Controls_AC\" : \"1\" , \"Controls_AF\" : \"0.500\" , \"Controls_AN\" : \"2\" , \"FS\" : 0 , \"GenotyperControls\" : \"Pindel\" , \"Group\" : \"Controls\" , \"HRun\" : 9 , \"SNPEFF_EFFECT\" : \"INTERGENIC\", \"SNPEFF_FUNCTIONAL_CLASS\" : \"NONE\", \"SNPEFF_IMPACT\" : \"MODIFIER\" }, \"_ident\" : \".\", \"_type\" : \"variant\", \"_landmark\" : \"1\", \"_refAllele\" : \"C\", \"_altAlleles\" : [  \"CT\" ], \"_minBP\" : 61871, \"_maxBP\" : 61871, \"samples\" : [ \t{ \t\"GT\" : \"0/1\", \t\"AD\" : 3, \t\"GenotypePositive\" : 1, \t\"sampleID\" : \"s_Mayo_TN_CC_757\" }], \"FORMAT\" : { \"max\" : { \"AD\" : 3 }, \"min\" : { \"AD\" : 3 }, \"GenotypePostitiveCount\" : 1, \"GenotypePositiveList\" : [  \"s_Mayo_TN_CC_757\" ] } }";
        String result = "{ \"CHROM\" : \"chr1\" , \"POS\" : \"61871\" , \"ID\" : \".\" , \"REF\" : \"C\" , \"ALT\" : \"CT\" , \"QUAL\" : \".\" , \"FILTER\" : \"PASS\" , \"INFO\" : { \"AC\" : [ 1] , \"AF\" : [ 0.5] , \"AN\" : 2 , \"Controls_AC\" : \"1\" , \"Controls_AF\" : \"0.500\" , \"Controls_AN\" : \"2\" , \"FS\" : 0 , \"GenotyperControls\" : \"Pindel\" , \"Group\" : \"Controls\" , \"HRun\" : 9 , \"SNPEFF_EFFECT\" : \"INTERGENIC\" , \"SNPEFF_FUNCTIONAL_CLASS\" : \"NONE\" , \"SNPEFF_IMPACT\" : \"MODIFIER\"} , \"_ident\" : \".\" , \"_type\" : \"variant\" , \"_landmark\" : \"1\" , \"_refAllele\" : \"C\" , \"_altAlleles\" : [ \"CT\"] , \"_minBP\" : 61871 , \"_maxBP\" : 61871 , \"FORMAT\" : { \"max\" : { \"AD\" : 3} , \"min\" : { \"AD\" : 3} , \"GenotypePostitiveCount\" : 1 , \"GenotypePositiveList\" : [ \"s_Mayo_TN_CC_757\"]}}";
        BasicDBObject bo = (BasicDBObject) JSON.parse(input);
        VCFParser parser = new VCFParser();
        assertEquals(result, bo.toString());
    }

    @Test
    public void testFixFormat(){
        String input = "{\"FORMAT\": {" +
                "\"PL\": 1," +
                "\"AD\": 1," +
                "\"GT\": 1," +
                "\"GQ\": 1," +
                "\"DP\": 1," +
                "\"MLPSAF\": 1," +
                "\"MLPSAC\": 1" +
                "}}";
        String result = "{ \"FORMAT\" : { \"min\" : { \"PL\" : 1 , \"AD\" : 1 , \"GT\" : 1 , \"GQ\" : 1 , \"DP\" : 1 , \"MLPSAF\" : 1 , \"MLPSAC\" : 1} , \"max\" : { \"PL\" : 1 , \"AD\" : 1 , \"GT\" : 1 , \"GQ\" : 1 , \"DP\" : 1 , \"MLPSAF\" : 1 , \"MLPSAC\" : 1}}}";
        BasicDBObject bo = (BasicDBObject) JSON.parse(input);
        VCFParser p = new VCFParser();
        assertEquals(result,p.fixFormat(bo).toString());

    }

    String json = "{ \"CHROM\" : \"chr1\" , \"POS\" : \"28218100\" , \"ID\" : \".\" , \"REF\" : \"T\" , \"ALT\" : \"C\" , \"QUAL\" : \"161.75\" , \"FILTER\" : \".\" , \"INFO\" : { \"AC\" : [ 39] , \"AF\" : [ 0.342] , \"AN\" : 114 , \"BaseQRankSum\" : -2.185 , \"DP\" : 22 , \"Dels\" : 0.0 , \"FS\" : 4.193 , \"HaplotypeScore\" : 0.0 , \"MLEAC\" : [ 37] , \"MLEAF\" : [ 0.325] , \"MQ\" : 70.0 , \"MQ0\" : 0 , \"MQRankSum\" : -0.282 , \"QD\" : 20.22 , \"ReadPosRankSum\" : -1.128} , \"_ident\" : \".\" , \"_type\" : \"variant\" , \"_landmark\" : \"1\" , \"_refAllele\" : \"T\" , \"_altAlleles\" : [ \"C\"] , \"_minBP\" : 28218100 , \"_maxBP\" : 28218100 , \"samples\" : [ { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 33.0] , \"maxPL\" : 33.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_319\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 39.0] , \"maxPL\" : 39.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_322\"} , { \"GT\" : \"1/1/1/1/1/1\" , \"AD\" : [ 0.0 , 1.0] , \"maxAD\" : 1.0 , \"minAD\" : 1.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 6.0 , \"MLPSAF\" : 1.0 , \"PL\" : [ 34.0 , 8.0 , 5.0 , 3.0 , 2.0 , 1.0 , 0.0] , \"maxPL\" : 34.0 , \"minPL\" : 0.0 , \"GenotypePositive\" : 1 , \"sampleID\" : \"s_Mayo_TN_CC_327\"} , { \"GT\" : \"0/0/0/1/1/1\" , \"AD\" : [ 1.0 , 1.0] , \"maxAD\" : 1.0 , \"minAD\" : 1.0 , \"DP\" : 2.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 3.0 , \"MLPSAF\" : 0.5 , \"PL\" : [ 30.0 , 3.0 , 1.0 , 0.0 , 1.0 , 3.0 , 30.0] , \"maxPL\" : 30.0 , \"minPL\" : 0.0 , \"GenotypePositive\" : 1 , \"sampleID\" : \"s_Mayo_TN_CC_335\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 37.0] , \"maxPL\" : 37.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_337\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 37.0] , \"maxPL\" : 37.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_339\"} , { \"GT\" : \"1/1/1/1/1/1\" , \"AD\" : [ 0.0 , 1.0] , \"maxAD\" : 1.0 , \"minAD\" : 1.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 6.0 , \"MLPSAF\" : 1.0 , \"PL\" : [ 37.0 , 8.0 , 5.0 , 3.0 , 2.0 , 1.0 , 0.0] , \"maxPL\" : 37.0 , \"minPL\" : 0.0 , \"GenotypePositive\" : 1 , \"sampleID\" : \"s_Mayo_TN_CC_348\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 38.0] , \"maxPL\" : 38.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_350\"} , { \"GT\" : \"1/1/1/1/1/1\" , \"AD\" : [ 0.0 , 1.0] , \"maxAD\" : 1.0 , \"minAD\" : 1.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 6.0 , \"MLPSAF\" : 1.0 , \"PL\" : [ 32.0 , 8.0 , 5.0 , 3.0 , 2.0 , 1.0 , 0.0] , \"maxPL\" : 32.0 , \"minPL\" : 0.0 , \"GenotypePositive\" : 1 , \"sampleID\" : \"s_Mayo_TN_CC_361\"} , { \"GT\" : \"1/1/1/1/1/1\" , \"AD\" : [ 0.0 , 1.0] , \"maxAD\" : 1.0 , \"minAD\" : 1.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 6.0 , \"MLPSAF\" : 1.0 , \"PL\" : [ 26.0 , 8.0 , 5.0 , 3.0 , 2.0 , 1.0 , 0.0] , \"maxPL\" : 26.0 , \"minPL\" : 0.0 , \"GenotypePositive\" : 1 , \"sampleID\" : \"s_Mayo_TN_CC_364\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 36.0] , \"maxPL\" : 36.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_367\"} , { \"GT\" : \"1/1/1/1/1/1\" , \"AD\" : [ 0.0 , 1.0] , \"maxAD\" : 1.0 , \"minAD\" : 1.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 6.0 , \"MLPSAF\" : 1.0 , \"PL\" : [ 30.0 , 8.0 , 5.0 , 3.0 , 2.0 , 1.0 , 0.0] , \"maxPL\" : 30.0 , \"minPL\" : 0.0 , \"GenotypePositive\" : 1 , \"sampleID\" : \"s_Mayo_TN_CC_374\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 37.0] , \"maxPL\" : 37.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_377\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 38.0] , \"maxPL\" : 38.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_380\"} , { \"GT\" : \"1/1/1/1/1/1\" , \"AD\" : [ 0.0 , 1.0] , \"maxAD\" : 1.0 , \"minAD\" : 1.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 6.0 , \"MLPSAF\" : 1.0 , \"PL\" : [ 37.0 , 8.0 , 5.0 , 3.0 , 2.0 , 1.0 , 0.0] , \"maxPL\" : 37.0 , \"minPL\" : 0.0 , \"GenotypePositive\" : 1 , \"sampleID\" : \"s_Mayo_TN_CC_382\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 36.0] , \"maxPL\" : 36.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_384\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 1.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 1.0 , \"GQ\" : 1.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 1.0 , 2.0 , 3.0 , 5.0 , 8.0 , 36.0] , \"maxPL\" : 36.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_389\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 2.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 2.0 , \"GQ\" : 2.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 2.0 , 4.0 , 6.0 , 10.0 , 16.0 , 77.0] , \"maxPL\" : 77.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_394\"} , { \"GT\" : \"0/0/0/0/0/0\" , \"AD\" : [ 2.0 , 0.0] , \"maxAD\" : 0.0 , \"minAD\" : 0.0 , \"DP\" : 2.0 , \"GQ\" : 2.0 , \"MLPSAC\" : 0.0 , \"MLPSAF\" : 0.0 , \"PL\" : [ 0.0 , 2.0 , 4.0 , 6.0 , 10.0 , 16.0 , 64.0] , \"maxPL\" : 64.0 , \"minPL\" : 0.0 , \"sampleID\" : \"s_Mayo_TN_CC_398\"}] , \"GenotypePostitiveCount\" : 7 , \"GenotypePositiveList\" : [ \"s_Mayo_TN_CC_327\" , \"s_Mayo_TN_CC_335\" , \"s_Mayo_TN_CC_348\" , \"s_Mayo_TN_CC_361\" , \"s_Mayo_TN_CC_364\" , \"s_Mayo_TN_CC_374\" , \"s_Mayo_TN_CC_382\"] , \"PLIntervalMin\" : 0.0 , \"PLIntervalMax\" : 77.0 , \"PLAverage\" : 8.661654135338345 , \"ADIntervalMin\" : 0.0 , \"ADIntervalMax\" : 1.0 , \"ADAverage\" : 0.3684210526315789}";
    String json2 = "{\"C.HROM\" : \"chr1\"}";
    String json3 = "{\"IN.FO\" : { \"A.C\" : [ 39] , \"AF\" : [ 0.342] , \"A.N\" : 114} }";
    String json4 = "{ \"CHROM\" : \"1\" , \"POS\" : \"215848808\" , \"ID\" : \"rs116645811\" , \"REF\" : \"G\" , \"ALT\" : \"A\" , \"QUAL\" : \".\" , \"FILTER\" : \".\" , \"INFO\" : { \".\" : true} , \"_ident\" : \"rs116645811\" , \"_type\" : \"variant\" , \"_landmark\" : \"1\" , \"_refAllele\" : \"G\" , \"_altAlleles\" : [ \"A\"] , \"_minBP\" : 215848808 , \"_maxBP\" : 215848808}";
    @Test
    public void testRemoveDots(){
        DBObject bo = (DBObject) JSON.parse(json);
        VCFParser parser = new VCFParser();
        parser.removeDots(bo, true);
        //make sure that if the json is complete, that nothing was removed...
        assertEquals(json, bo.toString());
        bo =  (DBObject) JSON.parse(json2);
        parser.removeDots(bo, true);
        assertEquals("{ \"C_HROM\" : \"chr1\"}", bo.toString());
        bo =  (DBObject) JSON.parse(json3);
        parser.removeDots(bo, true);
        assertEquals("{ \"IN_FO\" : { \"AF\" : [ 0.342] , \"A_C\" : [ 39] , \"A_N\" : 114}}", bo.toString());
        bo =  (DBObject) JSON.parse(json4);
        parser.removeDots(bo, true);
        assertEquals("{ \"CHROM\" : \"1\" , \"POS\" : \"215848808\" , \"ID\" : \"rs116645811\" , \"REF\" : \"G\" , \"ALT\" : \"A\" , \"QUAL\" : \".\" , \"FILTER\" : \".\" , \"INFO\" : { \"_\" : true} , \"_ident\" : \"rs116645811\" , \"_type\" : \"variant\" , \"_landmark\" : \"1\" , \"_refAllele\" : \"G\" , \"_altAlleles\" : [ \"A\"] , \"_minBP\" : 215848808 , \"_maxBP\" : 215848808}", bo.toString());

    }

    String json5 = "{ \"HEADER\" : { \"FORMAT\" : { \"PL\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\" , \"EntryType\" : \"FORMAT\"} , \"AD\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Allelic depths for the ref and alt alleles in the order listed\" , \"EntryType\" : \"FORMAT\"} , \"GT\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Genotype\" , \"EntryType\" : \"FORMAT\"} , \"GQ\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Genotype Quality\" , \"EntryType\" : \"FORMAT\"} , \"DP\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\" , \"EntryType\" : \"FORMAT\"} , \"MLPSAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Maximum likelihood expectation (MLE) for the alternate allele fraction, in the same order as listed, for each individual sample\" , \"EntryType\" : \"FORMAT\"} , \"MLPSAC\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Maximum likelihood expectation (MLE) for the alternate allele count, in the same order as listed, for each individual sample\" , \"EntryType\" : \"FORMAT\"}} , \"INFO\" : { \"bior.mt.mt\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"Controls_AN\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AN for Controls\" , \"EntryType\" : \"INFO\"} , \"PolyPhen.TERM\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The PolyPhen prediction.\" , \"EntryType\" : \"INFO\"} , \"HapMap.CEU_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Other allele frequency(Utah residents with Northern and Western European ancestry from the CEPH collection)\" , \"EntryType\" : \"INFO\"} , \"HapMap.YRI_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Other Allele Frequency(Han Chinese in Beijing China)\" , \"EntryType\" : \"INFO\"} , \"ESP6500.AFR_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"African American Minor Allele Frequency (in percent)\" , \"EntryType\" : \"INFO\"} , \"dbSNP.DiseaseVariant\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Variant disease name\" , \"EntryType\" : \"INFO\"} , \"1000Genomes.EUR_AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency for samples from EUR based on AC/AN\" , \"EntryType\" : \"INFO\"} , \"LOF\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Predicted loss of function effects for this variant. Format: 'Gene_Name | Gene_ID | Number_of_transcripts_in_gene | Percent_of_transcripts_affected' \" , \"EntryType\" : \"INFO\"} , \"InbreedingCoeff\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\" , \"EntryType\" : \"INFO\"} , \"Cases_AF\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AF for Cases\" , \"EntryType\" : \"INFO\"} , \"ReadPosRankSum\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\" , \"EntryType\" : \"INFO\"} , \"DP\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Approximate read depth; some reads may have been filtered\" , \"EntryType\" : \"INFO\"} , \"bior.phyloP.phyloP\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"DS\" : { \"number\" : 0 , \"type\" : \"Flag\" , \"Description\" : \"Were any of the samples downsampled?\" , \"EntryType\" : \"INFO\"} , \"bior.lrt.lrt\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"Controls_AF\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AF for Controls\" , \"EntryType\" : \"INFO\"} , \"Cases_AN\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AN for Cases\" , \"EntryType\" : \"INFO\"} , \"Controls_AC\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AC for Controls\" , \"EntryType\" : \"INFO\"} , \"STR\" : { \"number\" : 0 , \"type\" : \"Flag\" , \"Description\" : \"Variant is a short tandem repeat\" , \"EntryType\" : \"INFO\"} , \"BaseQRankSum\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\" , \"EntryType\" : \"INFO\"} , \"COSMIC.Mutation_AA\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The change that has occurred in the peptide sequence as a result of the mutation\" , \"EntryType\" : \"INFO\"} , \"QD\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Variant Confidence/Quality by Depth\" , \"EntryType\" : \"INFO\"} , \"bior.pp2.pp2\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"1000Genomes.ASN_AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency for samples from ASN based on AC/AN\" , \"EntryType\" : \"INFO\"} , \"MQ\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"RMS Mapping Quality\" , \"EntryType\" : \"INFO\"} , \"AC\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Allele count in genotypes, for each ALT allele, in the same order as listed\" , \"EntryType\" : \"INFO\"} , \"OMIM.Disease\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Phenotypes observed\" , \"EntryType\" : \"INFO\"} , \"SIFT.TERM\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The SIFT prediction.\" , \"EntryType\" : \"INFO\"} , \"Dels\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Fraction of Reads Containing Spanning Deletions\" , \"EntryType\" : \"INFO\"} , \"AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency, for each ALT allele, in the same order as listed\" , \"EntryType\" : \"INFO\"} , \"AN\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Total number of alleles in called genotypes\" , \"EntryType\" : \"INFO\"} , \"COSMIC.strand\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Strand based on GRCh37 genome build\" , \"EntryType\" : \"INFO\"} , \"HaplotypeScore\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Consistency of the site with at most two segregating haplotypes\" , \"EntryType\" : \"INFO\"} , \"PolyPhen.Score\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"The PolyPhen score.\" , \"EntryType\" : \"INFO\"} , \"set\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Source VCF for the merged record in CombineVariants\" , \"EntryType\" : \"INFO\"} , \"1000Genomes.AMR_AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency for samples from AMR based on AC/AN\" , \"EntryType\" : \"INFO\"} , \"NMD\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Predicted nonsense mediated decay effects for this variant. Format: 'Gene_Name | Gene_ID | Number_of_transcripts_in_gene | Percent_of_transcripts_affected' \" , \"EntryType\" : \"INFO\"} , \"MLEAC\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\" , \"EntryType\" : \"INFO\"} , \"MLEAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\" , \"EntryType\" : \"INFO\"} , \"FS\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Phred-scaled p-value using Fisher's exact test to detect strand bias\" , \"EntryType\" : \"INFO\"} , \"HapMap.JPT_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Other Allele Frequency(Japanese in Tokyo Japan)\" , \"EntryType\" : \"INFO\"} , \"COSMIC.Mutation_ID\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"This id is used to identify a mutation within the COSMIC database and is assigned as the mutation is curated\" , \"EntryType\" : \"INFO\"} , \"dbSNP.SuspectRegion\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Variant Suspect Reason Codes (may be more than one value added together) 0 - unspecified, 1 - Paralog, 2 - byEST, 4 - oldAlign, 8 - Para_EST, 16 - 1kg_failed, 1024 - other\" , \"EntryType\" : \"INFO\"} , \"Approved_Gene_Name\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The official gene name that has been approved by the HGNC and is publicly available\" , \"EntryType\" : \"INFO\"} , \"bior.sift.SIFTscore\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"ESP6500.EUR_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"European American Minor Allele Frequency (in percent)\" , \"EntryType\" : \"INFO\"} , \"SIFT.Score\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"The SIFT score.\" , \"EntryType\" : \"INFO\"} , \"Cases_AC\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AC for Cases\" , \"EntryType\" : \"INFO\"} , \"1000Genomes.AFR_AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency for samples from AFR based on AC/AN\" , \"EntryType\" : \"INFO\"} , \"dbSNP.ClinicalSig\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Variant Clinical Significance, 0 - unknown, 1 - untested, 2 - non-pathogenic, 3 - probable-non-pathogenic, 4 - probable-pathogenic, 5 - pathogenic, 6 - drug-response, 7 - histocompatibility, 255 - other\" , \"EntryType\" : \"INFO\"} , \"COSMIC.Mutation_CDS\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The change that has occurred in the nucleotide sequence as a result of the mutation\" , \"EntryType\" : \"INFO\"} , \"BGI200_Danish_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Estimated Minor Allele Frequency\" , \"EntryType\" : \"INFO\"} , \"MQRankSum\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\" , \"EntryType\" : \"INFO\"} , \"RU\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Tandem repeat unit (bases)\" , \"EntryType\" : \"INFO\"} , \"RPA\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Number of times tandem repeat unit is repeated, for each allele (including reference)\" , \"EntryType\" : \"INFO\"} , \"bior.gerp.gerp\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"MQ0\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Total Mapping Quality Zero Reads\" , \"EntryType\" : \"INFO\"} , \"HapMap.CHB_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Other Allele Frequency(Yoruban in Ibadan  Nigeria)\" , \"EntryType\" : \"INFO\"} , \"EFF\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Predicted effects for this variant.Format: 'Effect ( Effect_Impact | Functional_Class | Codon_Change | Amino_Acid_Change| Amino_Acid_length | Gene_Name | Transcript_BioType | Gene_Coding | Transcript_ID | Exon_Rank  | Genotype_Number [ | ERRORS | WARNINGS ] )' \" , \"EntryType\" : \"INFO\"} , \"UniprotID\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The UniProt identifier provided by the EBI.\" , \"EntryType\" : \"INFO\"}}} , \"SAMPLES\" : { \"s_Mayo_TN_CC_254\" : 189 , \"s_Mayo_TN_CC_255\" : 190 , \"s_Mayo_TN_CC_270\" : 207 , \"s_Mayo_TN_CC_277\" : 214 , \"s_Mayo_TN_CC_276\" : 213 , \"s_Mayo_TN_CC_275\" : 212 , \"s_Mayo_TN_CC_274\" : 211 , \"s_Mayo_TN_CC_278\" : 215 , \"s_Mayo_TN_CC_279\" : 216 , \"s_Mayo_TN_CC_556\" : 521 , \"s_Mayo_TN_CC_555\" : 520 , \"s_Mayo_TN_CC_554\" : 519 , \"s_Mayo_TN_CC_553\" : 518 , \"s_Mayo_TN_CC_750\" : 735 , \"s_Mayo_TN_CC_559\" : 524 , \"s_Mayo_TN_CC_558\" : 523 , \"s_Mayo_TN_CC_557\" : 522 , \"s_Mayo_TN_CC_740\" : 724 , \"s_Mayo_TN_CC_741\" : 725 , \"s_Mayo_TN_CC_742\" : 726 , \"s_Mayo_TN_CC_743\" : 727 , \"s_Mayo_TN_CC_744\" : 728 , \"s_Mayo_TN_CC_745\" : 729 , \"s_Mayo_TN_CC_540\" : 504 , \"s_Mayo_TN_CC_746\" : 730 , \"s_Mayo_TN_CC_541\" : 505 , \"s_Mayo_TN_CC_747\" : 731 , \"s_Mayo_TN_CC_260\" : 196 , \"s_Mayo_TN_CC_748\" : 732 , \"s_Mayo_TN_CC_749\" : 733 , \"s_Mayo_TN_CC_262\" : 198 , \"s_Mayo_TN_CC_261\" : 197 , \"s_Mayo_TN_CC_264\" : 200 , \"s_Mayo_TN_CC_263\" : 199 , \"s_Mayo_TN_CC_266\" : 202 , \"s_Mayo_TN_CC_265\" : 201 , \"s_Mayo_TN_CC_267\" : 203 , \"s_Mayo_TN_CC_268\" : 204 , \"s_Mayo_TN_CC_269\" : 205 , \"s_Mayo_TN_CC_543\" : 507 , \"s_Mayo_TN_CC_542\" : 506 , \"s_Mayo_TN_CC_545\" : 509 , \"s_Mayo_TN_CC_544\" : 508 , \"s_Mayo_TN_CC_547\" : 511 , \"s_Mayo_TN_CC_546\" : 510 , \"s_Mayo_TN_CC_549\" : 513 , \"s_Mayo_TN_CC_548\" : 512 , \"s_Mayo_TN_CC_738\" : 721 , \"s_Mayo_TN_CC_737\" : 720 , \"s_Mayo_TN_CC_739\" : 722 , \"s_Mayo_TN_CC_734\" : 717 , \"s_Mayo_TN_CC_733\" : 716 , \"s_Mayo_TN_CC_736\" : 719 , \"s_Mayo_TN_CC_735\" : 718 , \"s_Mayo_TN_CC_730\" : 713 , \"s_Mayo_TN_CC_732\" : 715 , \"s_Mayo_TN_CC_731\" : 714 , \"s_Mayo_TN_CC_729\" : 711 , \"s_Mayo_TN_CC_728\" : 710 , \"s_Mayo_TN_CC_727\" : 709 , \"s_Mayo_TN_CC_726\" : 708 , \"s_Mayo_TN_CC_725\" : 707 , \"s_Mayo_TN_CC_724\" : 706 , \"s_Mayo_TN_CC_723\" : 705 , \"s_Mayo_TN_CC_722\" : 704 , \"s_Mayo_TN_CC_721\" : 703 , \"s_Mayo_TN_CC_720\" : 702 , \"s_Mayo_TN_CC_716\" : 697 , \"s_Mayo_TN_CC_715\" : 696 , \"s_Mayo_TN_CC_718\" : 699 , \"s_Mayo_TN_CC_717\" : 698 , \"s_Mayo_TN_CC_719\" : 700 , \"s_Mayo_TN_CC_710\" : 691 , \"s_Mayo_TN_CC_712\" : 693 , \"s_Mayo_TN_CC_711\" : 692 , \"s_Mayo_TN_CC_714\" : 695 , \"s_Mayo_TN_CC_713\" : 694 , \"s_Mayo_TN_CC_707\" : 687 , \"s_Mayo_TN_CC_706\" : 686 , \"s_Mayo_TN_CC_705\" : 685 , \"s_Mayo_TN_CC_704\" : 684 , \"s_Mayo_TN_CC_709\" : 689 , \"s_Mayo_TN_CC_708\" : 688 , \"s_Mayo_TN_CC_703\" : 683 , \"s_Mayo_TN_CC_702\" : 682 , \"s_Mayo_TN_CC_701\" : 681 , \"s_Mayo_TN_CC_700\" : 680 , \"s_Mayo_TN_CC_588\" : 556 , \"s_Mayo_TN_CC_589\" : 557 , \"s_Mayo_TN_CC_586\" : 554 , \"s_Mayo_TN_CC_587\" : 555 , \"s_Mayo_TN_CC_585\" : 553 , \"s_Mayo_TN_CC_584\" : 552 , \"s_Mayo_TN_CC_583\" : 551 , \"s_Mayo_TN_CC_582\" : 550 , \"s_Mayo_TN_CC_581\" : 549 , \"s_Mayo_TN_CC_580\" : 548 , \"s_Mayo_TN_CC_597\" : 566 , \"s_Mayo_TN_CC_598\" : 567 , \"s_Mayo_TN_CC_599\" : 568 , \"s_Mayo_TN_CC_594\" : 563 , \"s_Mayo_TN_CC_593\" : 562 , \"s_Mayo_TN_CC_596\" : 565 , \"s_Mayo_TN_CC_595\" : 564 , \"s_Mayo_TN_CC_590\" : 559 , \"s_Mayo_TN_CC_592\" : 561 , \"s_Mayo_TN_CC_591\" : 560 , \"s_Mayo_TN_CC_203\" : 133 , \"s_Mayo_TN_CC_204\" : 134 , \"s_Mayo_TN_CC_201\" : 131 , \"s_Mayo_TN_CC_202\" : 132 , \"s_Mayo_TN_CC_207\" : 137 , \"s_Mayo_TN_CC_208\" : 138 , \"s_Mayo_TN_CC_205\" : 135 , \"s_Mayo_TN_CC_206\" : 136 , \"s_Mayo_TN_CC_209\" : 139 , \"s_Mayo_TN_CC_200\" : 130 , \"s_Mayo_TN_CC_212\" : 143 , \"s_Mayo_TN_CC_213\" : 144 , \"s_Mayo_TN_CC_214\" : 145 , \"s_Mayo_TN_CC_215\" : 146 , \"s_Mayo_TN_CC_216\" : 147 , \"s_Mayo_TN_CC_217\" : 148 , \"s_Mayo_TN_CC_218\" : 149 , \"s_Mayo_TN_CC_219\" : 150 , \"s_Mayo_TN_CC_211\" : 142 , \"s_Mayo_TN_CC_210\" : 141 , \"s_Mayo_TN_CC_37\" : 316 , \"s_Mayo_TN_CC_677\" : 654 , \"s_Mayo_TN_CC_36\" : 305 , \"s_Mayo_TN_CC_676\" : 653 , \"s_Mayo_TN_CC_35\" : 294 , \"s_Mayo_TN_CC_675\" : 652 , \"s_Mayo_TN_CC_34\" : 283 , \"s_Mayo_TN_CC_674\" : 651 , \"s_Mayo_TN_CC_33\" : 272 , \"s_Mayo_TN_CC_32\" : 261 , \"s_Mayo_TN_CC_31\" : 250 , \"s_Mayo_TN_CC_679\" : 656 , \"s_Mayo_TN_CC_30\" : 239 , \"s_Mayo_TN_CC_678\" : 655 , \"s_Mayo_TN_CC_159\" : 84 , \"s_Mayo_TN_CC_350\" : 295 , \"s_Mayo_TN_CC_157\" : 82 , \"s_Mayo_TN_CC_158\" : 83 , \"s_Mayo_TN_CC_353\" : 298 , \"s_Mayo_TN_CC_354\" : 299 , \"s_Mayo_TN_CC_39\" : 338 , \"s_Mayo_TN_CC_351\" : 296 , \"s_Mayo_TN_CC_38\" : 327 , \"s_Mayo_TN_CC_352\" : 297 , \"s_Mayo_TN_CC_358\" : 303 , \"s_Mayo_TN_CC_152\" : 77 , \"s_Mayo_TN_CC_357\" : 302 , \"s_Mayo_TN_CC_151\" : 76 , \"s_Mayo_TN_CC_356\" : 301 , \"s_Mayo_TN_CC_150\" : 75 , \"s_Mayo_TN_CC_355\" : 300 , \"s_Mayo_TN_CC_156\" : 81 , \"s_Mayo_TN_CC_155\" : 80 , \"s_Mayo_TN_CC_154\" : 79 , \"s_Mayo_TN_CC_359\" : 304 , \"s_Mayo_TN_CC_153\" : 78 , \"s_Mayo_TN_CC_40\" : 349 , \"s_Mayo_TN_CC_672\" : 649 , \"s_Mayo_TN_CC_673\" : 650 , \"s_Mayo_TN_CC_670\" : 647 , \"s_Mayo_TN_CC_671\" : 648 , \"s_Mayo_TN_CC_24\" : 173 , \"s_Mayo_TN_CC_664\" : 640 , \"s_Mayo_TN_CC_23\" : 162 , \"s_Mayo_TN_CC_663\" : 639 , \"s_Mayo_TN_CC_26\" : 195 , \"s_Mayo_TN_CC_666\" : 642 , \"s_Mayo_TN_CC_25\" : 184 , \"s_Mayo_TN_CC_665\" : 641 , \"s_Mayo_TN_CC_20\" : 129 , \"s_Mayo_TN_CC_668\" : 644 , \"s_Mayo_TN_CC_667\" : 643 , \"s_Mayo_TN_CC_22\" : 151 , \"s_Mayo_TN_CC_21\" : 140 , \"s_Mayo_TN_CC_669\" : 645 , \"s_Mayo_TN_CC_146\" : 70 , \"s_Mayo_TN_CC_147\" : 71 , \"s_Mayo_TN_CC_148\" : 72 , \"s_Mayo_TN_CC_149\" : 73 , \"s_Mayo_TN_CC_340\" : 284 , \"s_Mayo_TN_CC_28\" : 217 , \"s_Mayo_TN_CC_341\" : 285 , \"s_Mayo_TN_CC_27\" : 206 , \"s_Mayo_TN_CC_342\" : 286 , \"s_Mayo_TN_CC_343\" : 287 , \"s_Mayo_TN_CC_29\" : 228 , \"s_Mayo_TN_CC_345\" : 289 , \"s_Mayo_TN_CC_344\" : 288 , \"s_Mayo_TN_CC_347\" : 291 , \"s_Mayo_TN_CC_141\" : 65 , \"s_Mayo_TN_CC_346\" : 290 , \"s_Mayo_TN_CC_140\" : 64 , \"s_Mayo_TN_CC_349\" : 293 , \"s_Mayo_TN_CC_143\" : 67 , \"s_Mayo_TN_CC_348\" : 292 , \"s_Mayo_TN_CC_142\" : 66 , \"s_Mayo_TN_CC_145\" : 69 , \"s_Mayo_TN_CC_144\" : 68 , \"s_Mayo_TN_CC_660\" : 636 , \"s_Mayo_TN_CC_661\" : 637 , \"s_Mayo_TN_CC_662\" : 638 } , \"FORMAT\" : { \"min\" : { \"PL\" : 1 , \"AD\" : 1 , \"GT\" : 1 , \"GQ\" : 1 , \"DP\" : 1 , \"MLPSAF\" : 1 , \"MLPSAC\" : 1} , \"max\" : { \"PL\" : 1 , \"AD\" : 1 , \"GT\" : 1 , \"GQ\" : 1 , \"DP\" : 1 , \"MLPSAF\" : 1 , \"MLPSAC\" : 1}}}";
    @Test
    public void testRemoveDots2(){
        //this tests a real world case where the metadata is deeply nested and remove dots failed in the field...
        DBObject bo = (DBObject) JSON.parse(json5);
        VCFParser parser = new VCFParser();
        parser.removeDots(bo, true);
        String result = "{ \"HEADER\" : { \"FORMAT\" : { \"PL\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\" , \"EntryType\" : \"FORMAT\"} , \"AD\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Allelic depths for the ref and alt alleles in the order listed\" , \"EntryType\" : \"FORMAT\"} , \"GT\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Genotype\" , \"EntryType\" : \"FORMAT\"} , \"GQ\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Genotype Quality\" , \"EntryType\" : \"FORMAT\"} , \"DP\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\" , \"EntryType\" : \"FORMAT\"} , \"MLPSAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Maximum likelihood expectation (MLE) for the alternate allele fraction, in the same order as listed, for each individual sample\" , \"EntryType\" : \"FORMAT\"} , \"MLPSAC\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Maximum likelihood expectation (MLE) for the alternate allele count, in the same order as listed, for each individual sample\" , \"EntryType\" : \"FORMAT\"}} , \"INFO\" : { \"Controls_AN\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AN for Controls\" , \"EntryType\" : \"INFO\"} , \"LOF\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Predicted loss of function effects for this variant. Format: 'Gene_Name | Gene_ID | Number_of_transcripts_in_gene | Percent_of_transcripts_affected' \" , \"EntryType\" : \"INFO\"} , \"InbreedingCoeff\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\" , \"EntryType\" : \"INFO\"} , \"Cases_AF\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AF for Cases\" , \"EntryType\" : \"INFO\"} , \"ReadPosRankSum\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\" , \"EntryType\" : \"INFO\"} , \"DP\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Approximate read depth; some reads may have been filtered\" , \"EntryType\" : \"INFO\"} , \"DS\" : { \"number\" : 0 , \"type\" : \"Flag\" , \"Description\" : \"Were any of the samples downsampled?\" , \"EntryType\" : \"INFO\"} , \"Controls_AF\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AF for Controls\" , \"EntryType\" : \"INFO\"} , \"Cases_AN\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AN for Cases\" , \"EntryType\" : \"INFO\"} , \"Controls_AC\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AC for Controls\" , \"EntryType\" : \"INFO\"} , \"STR\" : { \"number\" : 0 , \"type\" : \"Flag\" , \"Description\" : \"Variant is a short tandem repeat\" , \"EntryType\" : \"INFO\"} , \"BaseQRankSum\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\" , \"EntryType\" : \"INFO\"} , \"QD\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Variant Confidence/Quality by Depth\" , \"EntryType\" : \"INFO\"} , \"MQ\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"RMS Mapping Quality\" , \"EntryType\" : \"INFO\"} , \"AC\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Allele count in genotypes, for each ALT allele, in the same order as listed\" , \"EntryType\" : \"INFO\"} , \"Dels\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Fraction of Reads Containing Spanning Deletions\" , \"EntryType\" : \"INFO\"} , \"AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency, for each ALT allele, in the same order as listed\" , \"EntryType\" : \"INFO\"} , \"AN\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Total number of alleles in called genotypes\" , \"EntryType\" : \"INFO\"} , \"HaplotypeScore\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Consistency of the site with at most two segregating haplotypes\" , \"EntryType\" : \"INFO\"} , \"set\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Source VCF for the merged record in CombineVariants\" , \"EntryType\" : \"INFO\"} , \"NMD\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Predicted nonsense mediated decay effects for this variant. Format: 'Gene_Name | Gene_ID | Number_of_transcripts_in_gene | Percent_of_transcripts_affected' \" , \"EntryType\" : \"INFO\"} , \"MLEAC\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\" , \"EntryType\" : \"INFO\"} , \"MLEAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\" , \"EntryType\" : \"INFO\"} , \"FS\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Phred-scaled p-value using Fisher's exact test to detect strand bias\" , \"EntryType\" : \"INFO\"} , \"Approved_Gene_Name\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The official gene name that has been approved by the HGNC and is publicly available\" , \"EntryType\" : \"INFO\"} , \"Cases_AC\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Original AC for Cases\" , \"EntryType\" : \"INFO\"} , \"BGI200_Danish_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Estimated Minor Allele Frequency\" , \"EntryType\" : \"INFO\"} , \"MQRankSum\" : { \"number\" : 1 , \"type\" : \"Float\" , \"Description\" : \"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\" , \"EntryType\" : \"INFO\"} , \"RU\" : { \"number\" : 1 , \"type\" : \"String\" , \"Description\" : \"Tandem repeat unit (bases)\" , \"EntryType\" : \"INFO\"} , \"RPA\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Number of times tandem repeat unit is repeated, for each allele (including reference)\" , \"EntryType\" : \"INFO\"} , \"MQ0\" : { \"number\" : 1 , \"type\" : \"Integer\" , \"Description\" : \"Total Mapping Quality Zero Reads\" , \"EntryType\" : \"INFO\"} , \"EFF\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Predicted effects for this variant.Format: 'Effect ( Effect_Impact | Functional_Class | Codon_Change | Amino_Acid_Change| Amino_Acid_length | Gene_Name | Transcript_BioType | Gene_Coding | Transcript_ID | Exon_Rank  | Genotype_Number [ | ERRORS | WARNINGS ] )' \" , \"EntryType\" : \"INFO\"} , \"UniprotID\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The UniProt identifier provided by the EBI.\" , \"EntryType\" : \"INFO\"} , \"bior_mt_mt\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"PolyPhen_TERM\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The PolyPhen prediction.\" , \"EntryType\" : \"INFO\"} , \"HapMap_CEU_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Other allele frequency(Utah residents with Northern and Western European ancestry from the CEPH collection)\" , \"EntryType\" : \"INFO\"} , \"HapMap_YRI_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Other Allele Frequency(Han Chinese in Beijing China)\" , \"EntryType\" : \"INFO\"} , \"ESP6500_AFR_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"African American Minor Allele Frequency (in percent)\" , \"EntryType\" : \"INFO\"} , \"dbSNP_DiseaseVariant\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Variant disease name\" , \"EntryType\" : \"INFO\"} , \"1000Genomes_EUR_AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency for samples from EUR based on AC/AN\" , \"EntryType\" : \"INFO\"} , \"bior_phyloP_phyloP\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"bior_lrt_lrt\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"COSMIC_Mutation_AA\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The change that has occurred in the peptide sequence as a result of the mutation\" , \"EntryType\" : \"INFO\"} , \"bior_pp2_pp2\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"1000Genomes_ASN_AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency for samples from ASN based on AC/AN\" , \"EntryType\" : \"INFO\"} , \"OMIM_Disease\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Phenotypes observed\" , \"EntryType\" : \"INFO\"} , \"SIFT_TERM\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The SIFT prediction.\" , \"EntryType\" : \"INFO\"} , \"COSMIC_strand\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Strand based on GRCh37 genome build\" , \"EntryType\" : \"INFO\"} , \"PolyPhen_Score\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"The PolyPhen score.\" , \"EntryType\" : \"INFO\"} , \"1000Genomes_AMR_AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency for samples from AMR based on AC/AN\" , \"EntryType\" : \"INFO\"} , \"HapMap_JPT_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Other Allele Frequency(Japanese in Tokyo Japan)\" , \"EntryType\" : \"INFO\"} , \"COSMIC_Mutation_ID\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"This id is used to identify a mutation within the COSMIC database and is assigned as the mutation is curated\" , \"EntryType\" : \"INFO\"} , \"dbSNP_SuspectRegion\" : { \"number\" : \".\" , \"type\" : \"Integer\" , \"Description\" : \"Variant Suspect Reason Codes (may be more than one value added together) 0 - unspecified, 1 - Paralog, 2 - byEST, 4 - oldAlign, 8 - Para_EST, 16 - 1kg_failed, 1024 - other\" , \"EntryType\" : \"INFO\"} , \"bior_sift_SIFTscore\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"ESP6500_EUR_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"European American Minor Allele Frequency (in percent)\" , \"EntryType\" : \"INFO\"} , \"SIFT_Score\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"The SIFT score.\" , \"EntryType\" : \"INFO\"} , \"1000Genomes_AFR_AF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Allele Frequency for samples from AFR based on AC/AN\" , \"EntryType\" : \"INFO\"} , \"dbSNP_ClinicalSig\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"Variant Clinical Significance, 0 - unknown, 1 - untested, 2 - non-pathogenic, 3 - probable-non-pathogenic, 4 - probable-pathogenic, 5 - pathogenic, 6 - drug-response, 7 - histocompatibility, 255 - other\" , \"EntryType\" : \"INFO\"} , \"COSMIC_Mutation_CDS\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"The change that has occurred in the nucleotide sequence as a result of the mutation\" , \"EntryType\" : \"INFO\"} , \"bior_gerp_gerp\" : { \"number\" : \".\" , \"type\" : \"String\" , \"Description\" : \"BioR property file missing description\" , \"EntryType\" : \"INFO\"} , \"HapMap_CHB_MAF\" : { \"number\" : \".\" , \"type\" : \"Float\" , \"Description\" : \"Other Allele Frequency(Yoruban in Ibadan  Nigeria)\" , \"EntryType\" : \"INFO\"}}} , \"SAMPLES\" : { \"s_Mayo_TN_CC_254\" : 189 , \"s_Mayo_TN_CC_255\" : 190 , \"s_Mayo_TN_CC_270\" : 207 , \"s_Mayo_TN_CC_277\" : 214 , \"s_Mayo_TN_CC_276\" : 213 , \"s_Mayo_TN_CC_275\" : 212 , \"s_Mayo_TN_CC_274\" : 211 , \"s_Mayo_TN_CC_278\" : 215 , \"s_Mayo_TN_CC_279\" : 216 , \"s_Mayo_TN_CC_556\" : 521 , \"s_Mayo_TN_CC_555\" : 520 , \"s_Mayo_TN_CC_554\" : 519 , \"s_Mayo_TN_CC_553\" : 518 , \"s_Mayo_TN_CC_750\" : 735 , \"s_Mayo_TN_CC_559\" : 524 , \"s_Mayo_TN_CC_558\" : 523 , \"s_Mayo_TN_CC_557\" : 522 , \"s_Mayo_TN_CC_740\" : 724 , \"s_Mayo_TN_CC_741\" : 725 , \"s_Mayo_TN_CC_742\" : 726 , \"s_Mayo_TN_CC_743\" : 727 , \"s_Mayo_TN_CC_744\" : 728 , \"s_Mayo_TN_CC_745\" : 729 , \"s_Mayo_TN_CC_540\" : 504 , \"s_Mayo_TN_CC_746\" : 730 , \"s_Mayo_TN_CC_541\" : 505 , \"s_Mayo_TN_CC_747\" : 731 , \"s_Mayo_TN_CC_260\" : 196 , \"s_Mayo_TN_CC_748\" : 732 , \"s_Mayo_TN_CC_749\" : 733 , \"s_Mayo_TN_CC_262\" : 198 , \"s_Mayo_TN_CC_261\" : 197 , \"s_Mayo_TN_CC_264\" : 200 , \"s_Mayo_TN_CC_263\" : 199 , \"s_Mayo_TN_CC_266\" : 202 , \"s_Mayo_TN_CC_265\" : 201 , \"s_Mayo_TN_CC_267\" : 203 , \"s_Mayo_TN_CC_268\" : 204 , \"s_Mayo_TN_CC_269\" : 205 , \"s_Mayo_TN_CC_543\" : 507 , \"s_Mayo_TN_CC_542\" : 506 , \"s_Mayo_TN_CC_545\" : 509 , \"s_Mayo_TN_CC_544\" : 508 , \"s_Mayo_TN_CC_547\" : 511 , \"s_Mayo_TN_CC_546\" : 510 , \"s_Mayo_TN_CC_549\" : 513 , \"s_Mayo_TN_CC_548\" : 512 , \"s_Mayo_TN_CC_738\" : 721 , \"s_Mayo_TN_CC_737\" : 720 , \"s_Mayo_TN_CC_739\" : 722 , \"s_Mayo_TN_CC_734\" : 717 , \"s_Mayo_TN_CC_733\" : 716 , \"s_Mayo_TN_CC_736\" : 719 , \"s_Mayo_TN_CC_735\" : 718 , \"s_Mayo_TN_CC_730\" : 713 , \"s_Mayo_TN_CC_732\" : 715 , \"s_Mayo_TN_CC_731\" : 714 , \"s_Mayo_TN_CC_729\" : 711 , \"s_Mayo_TN_CC_728\" : 710 , \"s_Mayo_TN_CC_727\" : 709 , \"s_Mayo_TN_CC_726\" : 708 , \"s_Mayo_TN_CC_725\" : 707 , \"s_Mayo_TN_CC_724\" : 706 , \"s_Mayo_TN_CC_723\" : 705 , \"s_Mayo_TN_CC_722\" : 704 , \"s_Mayo_TN_CC_721\" : 703 , \"s_Mayo_TN_CC_720\" : 702 , \"s_Mayo_TN_CC_716\" : 697 , \"s_Mayo_TN_CC_715\" : 696 , \"s_Mayo_TN_CC_718\" : 699 , \"s_Mayo_TN_CC_717\" : 698 , \"s_Mayo_TN_CC_719\" : 700 , \"s_Mayo_TN_CC_710\" : 691 , \"s_Mayo_TN_CC_712\" : 693 , \"s_Mayo_TN_CC_711\" : 692 , \"s_Mayo_TN_CC_714\" : 695 , \"s_Mayo_TN_CC_713\" : 694 , \"s_Mayo_TN_CC_707\" : 687 , \"s_Mayo_TN_CC_706\" : 686 , \"s_Mayo_TN_CC_705\" : 685 , \"s_Mayo_TN_CC_704\" : 684 , \"s_Mayo_TN_CC_709\" : 689 , \"s_Mayo_TN_CC_708\" : 688 , \"s_Mayo_TN_CC_703\" : 683 , \"s_Mayo_TN_CC_702\" : 682 , \"s_Mayo_TN_CC_701\" : 681 , \"s_Mayo_TN_CC_700\" : 680 , \"s_Mayo_TN_CC_588\" : 556 , \"s_Mayo_TN_CC_589\" : 557 , \"s_Mayo_TN_CC_586\" : 554 , \"s_Mayo_TN_CC_587\" : 555 , \"s_Mayo_TN_CC_585\" : 553 , \"s_Mayo_TN_CC_584\" : 552 , \"s_Mayo_TN_CC_583\" : 551 , \"s_Mayo_TN_CC_582\" : 550 , \"s_Mayo_TN_CC_581\" : 549 , \"s_Mayo_TN_CC_580\" : 548 , \"s_Mayo_TN_CC_597\" : 566 , \"s_Mayo_TN_CC_598\" : 567 , \"s_Mayo_TN_CC_599\" : 568 , \"s_Mayo_TN_CC_594\" : 563 , \"s_Mayo_TN_CC_593\" : 562 , \"s_Mayo_TN_CC_596\" : 565 , \"s_Mayo_TN_CC_595\" : 564 , \"s_Mayo_TN_CC_590\" : 559 , \"s_Mayo_TN_CC_592\" : 561 , \"s_Mayo_TN_CC_591\" : 560 , \"s_Mayo_TN_CC_203\" : 133 , \"s_Mayo_TN_CC_204\" : 134 , \"s_Mayo_TN_CC_201\" : 131 , \"s_Mayo_TN_CC_202\" : 132 , \"s_Mayo_TN_CC_207\" : 137 , \"s_Mayo_TN_CC_208\" : 138 , \"s_Mayo_TN_CC_205\" : 135 , \"s_Mayo_TN_CC_206\" : 136 , \"s_Mayo_TN_CC_209\" : 139 , \"s_Mayo_TN_CC_200\" : 130 , \"s_Mayo_TN_CC_212\" : 143 , \"s_Mayo_TN_CC_213\" : 144 , \"s_Mayo_TN_CC_214\" : 145 , \"s_Mayo_TN_CC_215\" : 146 , \"s_Mayo_TN_CC_216\" : 147 , \"s_Mayo_TN_CC_217\" : 148 , \"s_Mayo_TN_CC_218\" : 149 , \"s_Mayo_TN_CC_219\" : 150 , \"s_Mayo_TN_CC_211\" : 142 , \"s_Mayo_TN_CC_210\" : 141 , \"s_Mayo_TN_CC_37\" : 316 , \"s_Mayo_TN_CC_677\" : 654 , \"s_Mayo_TN_CC_36\" : 305 , \"s_Mayo_TN_CC_676\" : 653 , \"s_Mayo_TN_CC_35\" : 294 , \"s_Mayo_TN_CC_675\" : 652 , \"s_Mayo_TN_CC_34\" : 283 , \"s_Mayo_TN_CC_674\" : 651 , \"s_Mayo_TN_CC_33\" : 272 , \"s_Mayo_TN_CC_32\" : 261 , \"s_Mayo_TN_CC_31\" : 250 , \"s_Mayo_TN_CC_679\" : 656 , \"s_Mayo_TN_CC_30\" : 239 , \"s_Mayo_TN_CC_678\" : 655 , \"s_Mayo_TN_CC_159\" : 84 , \"s_Mayo_TN_CC_350\" : 295 , \"s_Mayo_TN_CC_157\" : 82 , \"s_Mayo_TN_CC_158\" : 83 , \"s_Mayo_TN_CC_353\" : 298 , \"s_Mayo_TN_CC_354\" : 299 , \"s_Mayo_TN_CC_39\" : 338 , \"s_Mayo_TN_CC_351\" : 296 , \"s_Mayo_TN_CC_38\" : 327 , \"s_Mayo_TN_CC_352\" : 297 , \"s_Mayo_TN_CC_358\" : 303 , \"s_Mayo_TN_CC_152\" : 77 , \"s_Mayo_TN_CC_357\" : 302 , \"s_Mayo_TN_CC_151\" : 76 , \"s_Mayo_TN_CC_356\" : 301 , \"s_Mayo_TN_CC_150\" : 75 , \"s_Mayo_TN_CC_355\" : 300 , \"s_Mayo_TN_CC_156\" : 81 , \"s_Mayo_TN_CC_155\" : 80 , \"s_Mayo_TN_CC_154\" : 79 , \"s_Mayo_TN_CC_359\" : 304 , \"s_Mayo_TN_CC_153\" : 78 , \"s_Mayo_TN_CC_40\" : 349 , \"s_Mayo_TN_CC_672\" : 649 , \"s_Mayo_TN_CC_673\" : 650 , \"s_Mayo_TN_CC_670\" : 647 , \"s_Mayo_TN_CC_671\" : 648 , \"s_Mayo_TN_CC_24\" : 173 , \"s_Mayo_TN_CC_664\" : 640 , \"s_Mayo_TN_CC_23\" : 162 , \"s_Mayo_TN_CC_663\" : 639 , \"s_Mayo_TN_CC_26\" : 195 , \"s_Mayo_TN_CC_666\" : 642 , \"s_Mayo_TN_CC_25\" : 184 , \"s_Mayo_TN_CC_665\" : 641 , \"s_Mayo_TN_CC_20\" : 129 , \"s_Mayo_TN_CC_668\" : 644 , \"s_Mayo_TN_CC_667\" : 643 , \"s_Mayo_TN_CC_22\" : 151 , \"s_Mayo_TN_CC_21\" : 140 , \"s_Mayo_TN_CC_669\" : 645 , \"s_Mayo_TN_CC_146\" : 70 , \"s_Mayo_TN_CC_147\" : 71 , \"s_Mayo_TN_CC_148\" : 72 , \"s_Mayo_TN_CC_149\" : 73 , \"s_Mayo_TN_CC_340\" : 284 , \"s_Mayo_TN_CC_28\" : 217 , \"s_Mayo_TN_CC_341\" : 285 , \"s_Mayo_TN_CC_27\" : 206 , \"s_Mayo_TN_CC_342\" : 286 , \"s_Mayo_TN_CC_343\" : 287 , \"s_Mayo_TN_CC_29\" : 228 , \"s_Mayo_TN_CC_345\" : 289 , \"s_Mayo_TN_CC_344\" : 288 , \"s_Mayo_TN_CC_347\" : 291 , \"s_Mayo_TN_CC_141\" : 65 , \"s_Mayo_TN_CC_346\" : 290 , \"s_Mayo_TN_CC_140\" : 64 , \"s_Mayo_TN_CC_349\" : 293 , \"s_Mayo_TN_CC_143\" : 67 , \"s_Mayo_TN_CC_348\" : 292 , \"s_Mayo_TN_CC_142\" : 66 , \"s_Mayo_TN_CC_145\" : 69 , \"s_Mayo_TN_CC_144\" : 68 , \"s_Mayo_TN_CC_660\" : 636 , \"s_Mayo_TN_CC_661\" : 637 , \"s_Mayo_TN_CC_662\" : 638} , \"FORMAT\" : { \"min\" : { \"PL\" : 1 , \"AD\" : 1 , \"GT\" : 1 , \"GQ\" : 1 , \"DP\" : 1 , \"MLPSAF\" : 1 , \"MLPSAC\" : 1} , \"max\" : { \"PL\" : 1 , \"AD\" : 1 , \"GT\" : 1 , \"GQ\" : 1 , \"DP\" : 1 , \"MLPSAF\" : 1 , \"MLPSAC\" : 1}}}";
        assertEquals(result,bo.toString());
    }


    @Test
    public void testParseBadFile(){
        boolean reporting = true;
        String junkfile = "src/test/resources/testData/foo.tsv";
        VCFParser parser = new VCFParser();
        parser.setTypeAhead(new TypeAheadCollection());
        parser.setReporting(reporting);
        parser.setTesting(true); //don't try to place it in mongo
        try {
            parser.parse(junkfile,"wSOMERANDOMWORKSPACE");
            fail();
        } catch (Exception e) {//InvalidPipeInputException
            //e.printStackTrace();  //On success should get here
        }
    }

    @Test
    public void testInfinityDoubleisSameAsConfig(){
        //tests that +/- INFINITY and other parameters are gotten from the sys.properties file and not from java
        String infile = "src/test/resources/testData/1KG.chr22.anno.first1500.vcf";
        String workspace = "foowork";
        VCFParser parser = new VCFParser();
        Sender sender = new FileSender(parser.getErrorFile(workspace));
        VCF2VariantPipe vcf = new VCF2VariantPipe(sender, true, false);
        Pipe p = parser.getPipeline(vcf, infile);
        while(p.hasNext()){
            History h = (History) p.next();
            String json = h.get(h.size() - 1);
            if(json.contains("16053863")) {
                System.out.println(json);
                //TODO: drill into the resulting json to ensure we have: "CUSTOM":{"max":{"AD":1.0E-307},"min":{"AD":1.0E308
                DBObject result = (DBObject) JSON.parse(json);
                Double maxAD =  (Double) ( (((DBObject)((DBObject)result.get("CUSTOM")).get("max")).get("AD")) );
                Double minAD =  (Double) ( (((DBObject)((DBObject)result.get("CUSTOM")).get("min")).get("AD")) );
                assertEquals(1.0E-307, maxAD,0.01 );
                assertEquals(1.0E308, minAD,0.01 );
                //NOTE: if this test does not pass, you need to put the following in your sys.properties: (no forward slashes of course)
                //#Values for INFINITY, 1/INFINITY, NotANumber ect that work in both the javascript importer AND java.
                //Double_MAX_VALUE = 1.0e308
                //Double_MIN_VALUE = 1.0e-307
                //Double_NEGATIVE_INFINITY = -1.0e307
                //NaN_KEY = NaN
                break;
            }
        }
    }

}
